qiskit/pulse/builder.py

```diff
@@ -286,62 +294,81 @@ def __init__(self,
         self._backend_ctx_token = None
 
-        #: pulse.Schedule: Active schedule of BuilderContext.
-        self._context_schedule = None
-
         #: QuantumCircuit: Lazily constructed quantum circuit
         self._lazy_circuit = None
 
-        # ContextManager: Default alignment context.
-        self._default_alignment_context = _align(default_alignment)
-
-        self._transpiler_settings = default_transpiler_settings or {}
-
-        if default_circuit_scheduler_settings is None:
-            default_circuit_scheduler_settings = {}
-        self._circuit_scheduler_settings = \
-            default_circuit_scheduler_settings or {}
-
-        # pulse.Schedule: Root program context-schedule
-        self._schedule = schedule or Schedule(name=name)
-
-        self.set_context_schedule(Schedule())
-
-    def __enter__(self) -> Schedule:
+        #: Dict[str, Any]: Transpiler setting dictionary.
+        self._transpiler_settings = default_transpiler_settings or dict()
+
+        #: Dict[str, Any]: Scheduler setting dictionary.
+        self._circuit_scheduler_settings = default_circuit_scheduler_settings or dict()
+
+        #: List[ScheduleBlock]: Stack of context.
+        self._context_stack = []
+
+        #: str: Name of the output program
+        self._name = name
+
+        # Add root block if provided. Schedule will be built on top of this.
+        if block is not None:
+            if isinstance(block, ScheduleBlock):
+                root_block = block
+            elif isinstance(block, Schedule):
+                root_block = ScheduleBlock()
+                root_block.append(instructions.Call(subroutine=block))
+            else:
+                raise exceptions.PulseError(f'Input `block` type {block.__class__.__name__} is '
+                                            'not a valid format. Specify a pulse program.')
+            self._context_stack.append(root_block)
+
+        # Set default alignment context
+        alignment = _PulseBuilder.__alignment_kinds__.get(default_alignment, default_alignment)
+        if not isinstance(alignment, AlignmentKind):
+            raise exceptions.PulseError(f'Given `default_alignment` {repr(default_alignment)} is '
+                                        'not a valid transformation. Set one of '
+                                        f'{", ".join(_PulseBuilder.__alignment_kinds__.keys())}, '
+                                        'or set an instance of `AlignmentKind` subclass.')
+        self.push_context(alignment)
+
+    def __enter__(self) -> ScheduleBlock:
         """Enter this builder context and yield either the supplied schedule
         or the schedule created for the user.
 
         Returns:
             The schedule that the builder will build on.
         """
         self._backend_ctx_token = BUILDER_CONTEXTVAR.set(self)
-        self._default_alignment_context.__enter__()
-        return self._schedule
+        output = self._context_stack[0]
```

Comment:
`get_context` to get current context is added. However this requires the root context, rather then the current `self._context_stack[-1]`. This is kept as-is since this is only place requiring the root one.